### DISCUSSION TITLE
ipsec.conf working, vici doesn't

### DISCUSSION THREAD

**Discussion by misch42** [strongswan/strongswan#1405] [General]
Hi,

I have a ipsec.conf based config, the connection starts. The logs show:
07[CFG] received stroke: initiate 'myconn'
07[IKE] initiating IKE_SA myconn[1] to x.x.x.x
07[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
07[NET] sending packet: from y.y.y.y[500] to x.x.x.x[500] (1012 bytes)
09[CFG] received stroke: add connection 'sa_0'
09[CFG] added child to existing configuration 'myconn'
09[CFG] received stroke: initiate 'sa_0'
09[CFG] received stroke: add connection 'sa_1'
09[CFG] added child to existing configuration 'myconn'
09[CFG] received stroke: initiate 'sa_1'
05[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (423 bytes)
05[ENC] parsed IKE_SA_INIT response 0 [ SA KE No V V CERTREQ N(FRAG_SUP) V ]
05[IKE] received Cisco Delete Reason vendor ID
05[IKE] received Cisco Copyright (c) 2009 vendor ID
05[IKE] received FRAGMENTATION vendor ID
05[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384
05[IKE] received 3 cert requests for an unknown ca
05[IKE] authentication of 'y.y.y.y' (myself) with pre-shared key
05[IKE] establishing CHILD_SA myconn{1}
05[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
05[NET] sending packet: from y.y.y.y[500] to x.x.x.x[500] (416 bytes)
06[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (160 bytes)
06[ENC] parsed IKE_AUTH response 1 [ V IDr AUTH N(NO_PROP) ]
06[IKE] authentication of 'x.x.x.x' with pre-shared key successful
06[IKE] IKE_SA myconn[1] established between y.y.y.y[y.y.y.y]...x.x.x.x[x.x.x.x]
06[IKE] scheduling reauthentication in 2756s
06[IKE] maximum IKE_SA lifetime 3296s
06[IKE] received NO_PROPOSAL_CHOSEN notify, no CHILD_SA built
06[IKE] failed to establish CHILD_SA, keeping IKE_SA
06[IKE] establishing CHILD_SA sa_0{2}
06[ENC] generating CREATE_CHILD_SA request 2 [ SA No KE TSi TSr ]
06[NET] sending packet: from y.y.y.y[500] to x.x.x.x[500] (464 bytes)
08[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (352 bytes)
08[ENC] parsed CREATE_CHILD_SA response 2 [ SA No KE TSi TSr ]
08[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/ECP_384/NO_EXT_SEQ
08[IKE] CHILD_SA sa_0{2} established with SPIs cdd0c810_i 5558ed6a_o and TS X.X.X.X/24 === Y.Y.Y.Y/28
08[IKE] establishing CHILD_SA sa_1{3}
08[ENC] generating CREATE_CHILD_SA request 3 [ SA No KE TSi TSr ]
08[NET] sending packet: from y.y.y.y[500] to x.x.x.x[500] (464 bytes)
11[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (352 bytes)
11[ENC] parsed CREATE_CHILD_SA response 3 [ SA No KE TSi TSr ]
11[CFG] selected proposal: ESP:AES_CBC_256/HMAC_SHA2_256_128/ECP_384/NO_EXT_SEQ
11[IKE] CHILD_SA sa_1{3} established with SPIs c812a3f5_i 2d439ac7_o and TS X.X.X.X/24 === Y.Y.Y.Y/27

Seems a bit strange, since we receive a NO_PROPOSAL_CHOSEN, but the ipsec based config starts the child connection anyway.

If I use a swanctl based config, the child  connection will not  start. The logs say:

05[CFG] vici initiate CHILD_SA 'myconn-net'
12[IKE] initiating IKE_SA myconn[3] to x.x.x.x
12[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
12[NET] sending packet: from y.y.y.y[500] to x.x.x.x[500] (304 bytes)
06[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (423 bytes)
06[ENC] parsed IKE_SA_INIT response 0 [ SA KE No V V CERTREQ N(FRAG_SUP) V ]
06[IKE] received Cisco Delete Reason vendor ID
06[IKE] received Cisco Copyright (c) 2009 vendor ID
06[IKE] received FRAGMENTATION vendor ID
06[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_384
06[IKE] received 3 cert requests for an unknown ca
06[IKE] authentication of 'y.y.y.y' (myself) with pre-shared key
06[IKE] establishing CHILD_SA myconn-net{3}
06[ENC] generating IKE_AUTH request 1 [ IDi N(INIT_CONTACT) IDr AUTH SA TSi TSr N(MOBIKE_SUP) N(ADD_4_ADDR) N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
06[NET] sending packet: from y.y.y.y[500] to x.x.x.x[500] (272 bytes)
13[NET] received packet: from x.x.x.x[500] to y.y.y.y[500] (160 bytes)
13[ENC] parsed IKE_AUTH response 1 [ V IDr AUTH N(NO_PROP) ]
13[IKE] authentication of 'x.x.x.x' with pre-shared key successful
13[IKE] IKE_SA myconn[3] established between y.y.y.y[y.y.y.y]...x.x.x.x[x.x.x.x]
13[IKE] scheduling reauthentication in 26634s
13[IKE] maximum IKE_SA lifetime 29514s
13[IKE] received NO_PROPOSAL_CHOSEN notify, no CHILD_SA built
13[IKE] failed to establish CHILD_SA, keeping IKE_SA

strongswan stops after receiving the NO_PROPOSAL_CHOSEN, and does not start the children after that.

Any idea how to configure swanctl.conf  to bring up the children?

Interesting enough, an "ipsec up myconn" also fails bringing up the children.

Michael

**Comment by misch42** [comment]
Additional:

I found https://wiki.strongswan.org/issues/3281 describing quite the same problem. So I added the "!" to the end of my esp_proposals:

children {
  myconn-net {
    ...
    esp_proposals = aes256-sha256-ecp384!
  }
}

Now the swanctl --load-all throws an error:

loading connection 'myconn' failed: invalid value for: esp_proposals, config discarded

Ideas?

Michael

**Reply by Thermi** [reply]
swanctl does not add any default proposals by default so the ! is not supported because that's the default behavior already.
You can achieve the same behavior as with ipsec.conf by appending, default.

Can you please provide your complete ipsec.conf and swanctl.conf you are using?

ipsec.conf:
```
ike=aes128-sha1-modp2048!
```
swanctl.conf
```
proposals = aes128-sah1-modp2048
```
----

ipsec.conf
```
ike=aes128-sha1-modp2048
```
swanctl.conf
```
proposals=aes128-sha1-modp2048,default
```

**Comment by misch42** [comment]
ipsec.conf:

config setup
        charondebug="all"
        uniqueids=yes

conn myconn
        type=tunnel
        auto=start
        keyexchange=ikev2
        authby=secret
        leftid=y.y.y.y
        leftsubnet=Y.Y.Y.Y/24
        right=x.x.x.x
        ike=aes256-sha2_256-ecp384
        esp=aes256-sha2_256-ecp384
        aggressive=no
        keyingtries=%forever
        ikelifetime=3600s
        lifetime=86400s
        dpddelay=30s
        dpdtimeout=120s
        dpdaction=restart


conn myconn-net
                rightsubnet=X.X.X.X/28
                also=myconn
                auto=start

swanctl.conf:

connections {

   myconn {
      local_addrs  = y.y.y.y
      remote_addrs = x.x.x.x
      dpd_delay = 30s
      proposals = aes256-sha256-ecp384
      keyingtries = 0

      local {
         auth = psk
         id = y.y.y.y
      }
      remote {
         auth = psk
         id = x.x.x.x 
      }

      children {
         myconn-net {
            local_ts = Y.Y.Y.Y/28
            remote_ts = X.X.X.X/24
            rekey_time = 3600
            esp_proposals=aes256-sha256-ecp384
         }
      }
      version = 2
      mobike = yes  # not needed.
      reauth_time = 28800
   }
}

**Comment by Thermi** [comment]
Try appending ",default" to the proposals so the default is added, or ask the other peer to initiate. Then you can see their proposals.

Am 9. November 2022 08:41:18 UTC schrieb Michael Schwartzkopff ***@***.***>:
>ipsec.conf:
>
>config setup
>        charondebug="all"
>        uniqueids=yes
>
>conn myconn
>        type=tunnel
>        auto=start
>        keyexchange=ikev2
>        authby=secret
>        leftid=y.y.y.y
>        leftsubnet=Y.Y.Y.Y/24
>        right=x.x.x.x
>        ike=aes256-sha2_256-ecp384
>        esp=aes256-sha2_256-ecp384
>        aggressive=no
>        keyingtries=%forever
>        ikelifetime=3600s
>        lifetime=86400s
>        dpddelay=30s
>        dpdtimeout=120s
>        dpdaction=restart
>
>
>conn myconn-net
>                rightsubnet=X.X.X.X/28
>                also=myconn
>                auto=start
>
>swanctl.conf:
>
>connections {
>
>   myconn {
>      local_addrs  = y.y.y.y
>      remote_addrs = x.x.x.x
>      dpd_delay = 30s
>      proposals = aes256-sha256-ecp384
>      keyingtries = 0
>
>      local {
>         auth = psk
>         id = y.y.y.y
>      }
>      remote {
>         auth = psk
>         id = x.x.x.x 
>      }
>
>      children {
>         myconn-net {
>            local_ts = Y.Y.Y.Y/28
>            remote_ts = X.X.X.X/24
>            rekey_time = 3600
>            esp_proposals=aes256-sha256-ecp384
>         }
>      }
>      version = 2
>      mobike = yes  # not needed.
>      reauth_time = 28800
>   }
>}
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/strongswan/strongswan/discussions/1405#discussioncomment-4094711
>You are receiving this because you commented.
>
>Message ID: ***@***.***>
-- 
Sent from mobile. Please excuse any typos or brevity

**Comment by misch42** [comment]
Added a ,default to swanctl,conf. No change in behaviour.

I think the problem ist that a vici-based config gives up after receiving the "NO_PROPOSAL_CHOSEN" message, while the ipsec.conf based config goes on trying. Perhaps this is because I shortened the config a litte bit. In reality I have two children in ipsec.conf, and only one (for testing) in swanctl.conf.

**Reply by tobiasbrunner** [reply]
> Perhaps this is because I shortened the config a litte bit. In reality I have two children in ipsec.conf, and only one (for testing) in swanctl.conf.

You actually have three in ipsec.conf, at least according to the log, `myconn`, which fails like it does for swanctl.conf, and `sa_0` and `sa_1`, which apparently work. Please post the complete configs.

**Comment by misch42** [comment]
No success including two child sections in the swanctl.conf.
